File: solutions/0011-container-with-most-water/solution.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int start_index = 0, end_index =(height.size() - 1);
        int prev_area_max = 0, cur_area = 0, cur_area_max = -1;
        while (start_index < end_index) {
            cur_area = (std::min(height[start_index], height[end_index]) *
                        (end_index - start_index));
            prev_area_max = cur_area_max;
            cur_area_max = std::max(cur_area, prev_area_max);
            /* std::cout << "{" << height[start_index] << "," << start_index << "}" <<
                      "| {" << height[end_index] << "," << end_index << "} = " << cur_area << std::endl; */
            // Move the index based on what is smaller
            if (height[start_index] < height[end_index]) {
                ++start_index;
            } else {
                --end_index;
            }
        }
        // std::cout << "cur_area_max = " << cur_area_max << std::endl;
        return cur_area_max;
    }
};

File: solutions/0011-container-with-most-water/solution.go
func maxArea(height []int) int {
    start, end := 0, (len(height) - 1)
    prevAreaMax, curArea, curAreaMax := 0, 0, 0

    // Lets move the indices after area calculations
    // Two pointers approach. Earlier the better as bottom needs to be maximized
    for start < end {
        curArea = min(height[start], height[end]) * (end - start)
        prevAreaMax = curAreaMax
        curAreaMax = max(curArea, prevAreaMax)
        if height[start] < height[end] {
            start++
        } else {
            end--
        }
    }
    return curAreaMax
}
